### PR TITLE
TRITON-2464 node-kstat leaks nvlists

### PR DESCRIPTION
**Pull request by danmcd**
Needs more testing, but putting it up here now for review!

Biggest concern I have is that freeing these nvlists (as they should be freed) might kneecap something else that makes bad assumptions about nvlist content.

### CODE REVIEW

**Review by bahamat — CHANGES_REQUESTED** [review]
copyright update

File: kstat.c [hunk 1]
```diff
@@ -159,23 +159,27 @@ kstatjs_parsearg_int(const nvlist_t *arg, const char *str, int *p)
...
+	    kstatjs_parsearg_int(args, "instance", instancep) != 0) {
+		rc = -1;
 	}
 
-	if (kstatjs_parsearg_int(args, "instance", instancep) != 0)
-		return (-1);
-
-	return (0);
+	/* No caller uses "args" after it's done here. */
+	nvlist_free(args);
```
Comment (jperkin) [inline kstat.c:181]: This introduces a new warning:

```
../kstat.c: In function 'kstatjs_parsearg':
../kstat.c:181:21: warning: passing argument 1 of 'nvlist_free' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
  181 |         nvlist_free(args);
      |                     ^~~~
In file included from /usr/include/libnvpair.h:29,
                 from ../addon.h:13,
                 from ../kstat.c:14:
/usr/include/sys/nvpair.h:156:18: note: expected 'nvlist_t *' {aka 'struct nvlist *'} but argument is of type 'const nvlist_t *' {aka 'const struct nvlist *'}
  156 | void nvlist_free(nvlist_t *);
      |                  ^~~~~~~~~~
```

**Comment by jperkin** [comment]
Ok, so the problem is when we're trying to read a kstat that doesn't exist, which explains why all my tests didn't show this as all of the examples and test cases we have read legitimate data.

Here's a minimal reproducer:

```node
var kstat = require('bindings')('kstat');
var reader = new kstat.Reader({ 'class': 'does-not-exist' } );
while (1) {
        reader.read();
}
```

Without the fix this quickly balloons:

```
  9872 root     1938M 1840M cpu10    1    0   0:00:34 2.5% node/8
```

and shows leaks:

```
> ::findleaks -d
CACHE             LEAKED           BUFFER CALLER
0000000004e44028 16129356 000000000504ad50 ?
0000000004e48028 16090863 0000000004f26150 ?
------------------------------------------------------------------------
           Total 32220219 buffers, 2061478128 bytes

umem_alloc_48 leak: 16129356 buffers, 48 bytes each, 774209088 bytes total,
    sample addr 504ad50
umem_alloc_80 leak: 16090863 buffers, 80 bytes each, 1287269040 bytes total,
    sample addr 4f26150
```

With the fix applied, apart from the minor warning, it looks great:

```
 91439 root      140M   42M cpu1     1    0   0:01:04 3.0% node/8
```

```
> ::findleaks -d
findleaks: no memory leaks detected
```

Thanks Dan. Sorry I missed this the first time around. I don't know if we also want to look at cmon-agent and prune the requests for kstats that don't exist? May be wasting time and resources, regardless of the memory leak.

FWIW I ran the test suite and the examples and all of them seem to work as expected, so I don't anticipate any regressions here. Hopefully @bahamat can also verify that all of the cmon data with this fix still looks good.

File: kstat.c (hunk 1 above)
Comment (danmcd) [reply kstat.c:181]: Thank you... fixing by losing "const" from `kstatjs_parsearg()` (but keeping the `const` in the ones it calls).

**Review by jperkin — DISMISSED** [review]

**Review by bahamat — APPROVED** [review]

**Pull request merged**